subiquity/controllers/refresh.py

```diff
@@ -47,11 +54,93 @@ def __init__(self, common):
         super().__init__(common)
         self.snap_name = os.environ.get("SNAP_NAME", "subiquity")
         self.check_state = CheckState.NOT_STARTED
+        self.switch_state = SwitchState.NOT_STARTED
+        self.network_state = "down"
         self.view = None
         self.offered_first_time = False
         self.answers = self.all_answers.get("Refresh", {})
 
+    def start(self):
+        self.switch_state = SwitchState.SWITCHING
+        channel = self.get_refresh_channel()
+        self.run_in_bg(
+            lambda: self._bg_switch_snap(channel),
+            self._snap_switched)
+
+    def get_refresh_channel(self):
+        """Return the channel we should refresh subiquity to."""
+
+        with open('/proc/cmdline') as fp:
+            cmdline = fp.read()
+        prefix = "subquity-channel="
+        for arg in cmdline.split():
+            if arg.startswith(prefix):
+                log.debug("found cmdline arg %s", arg)
+                return arg[len(prefix):]
+
+        info_file = '/cdrom/.disk/info'
+        try:
+            fp = open(info_file)
+        except FileNotFoundError:
+            if self.opts.dry_run:
+                info = (
+                    'Ubuntu-Server 18.04.2 LTS "Bionic Beaver" - '
+                    'Release amd64 (20190214.3)')
+            else:
+                log.debug("failed to find .disk/info file")
+                return
+        else:
+            with fp:
+                info = fp.read()
+        release = info.split()[1]
```

Comment:
I don't like any of this. Daily builds might not have digits in the info file at all. And 18.04.2 is too of a restrictive track, no? We probably want just the 18.04 track, no? Imho you should simply inspect `/etc/os-release` the `VERSION_ID="19.04"` field.

Reply:
I can't see how a daily build could _not_  have digits in the info file, looking at debian-cd, although perhaps I should be a bit more defensive here. And no, 18.04.2 is not too restrictive, it is quite possible for there to be changes that we don't want to go back to a previous release -- for example if we make a change that depends on an addition to the server-ship-live seed between 18.04.3 and 18.04.4 we might not want users of 18.04.3 media to get that as it would break offline installs.

Reply:
> for example if we make a change that depends on an addition to the server-ship-live seed between 18.04.3 and 18.04.4 we might not want users of 18.04.3 media to get that as it would break offline installs.

Haha this example makes no sense does it. Time for bed.

But the point is that the most likely reason to not want to offer a new subiquity to a user of an older release is that new subiquity depends on some change in how the image is made. Noone has yet suggested a single (non arch specific) feature for subiquity that we would not want to offer to 18.04 users (really, I went through all bugs and cards to check this)

Reply:
ok, that's fine. It's just for each point release, we'd then need to open and close a track. And i guess by default we will make the fallback chain to be "stable/18.04.3 -> stable/18.04 -> stable" or maybe simply just "stable/18.04.3 -> stable"

Reply:
We don't have support for creating a non-default fallback chain for closed tracks.  It will be "stable/18.04.3 -> stable"; and if at any point we reopen stable/18.04, we will also need to copy to stable/18.04.{3,4} etc.